akka-sample-sharding-scala/killrweather/src/main/scala/sample/killrweather/WeatherStation.scala

```diff
@@ -0,0 +1,132 @@
+ * but becomes useful when you have a collection of 'heavy' actors (in terms of processing or state)
+ * so you need to distribute them across several nodes.
+ *
+ * A sharded `WeatherStation` has a set of recorded datapoints
+ * For each weather station common cumulative computations can be run:
+ * aggregate, averages, high/low, topK (e.g. the top N highest temperatures).
+ *
+ * Note that since this station is not persistent, if Akka Cluster Sharding rebalances it - moves it to another
+ * node because of cluster nodes added removed etc - it will loose all its state.
+ */
+private[killrweather] object WeatherStation {
+
+  // setup for using WeatherStations through Akka Cluster Sharding
+  // these could also live elsewhere and the WeatherStation class be completely
+  // oblivious to being used in sharding
+  val TypeKey: EntityTypeKey[WeatherStation.Command] =
+    EntityTypeKey[WeatherStation.Command]("WeatherStation")
+
+  def initSharding(system: ActorSystem[_]): Unit =
+    ClusterSharding(system).init(Entity(TypeKey) { entityContext =>
+      WeatherStation(entityContext.entityId)
+    })
+
+
+  // actor commands and responses
+  sealed trait Command extends CborSerializable
+
+  final case class Record(data: Data, processingTimestamp: Long, replyTo: ActorRef[DataRecorded]) extends Command
+  final case class DataRecorded(wsid: String) extends CborSerializable
+
+  final case class Query(wsid: String, dataType: DataType, func: Function, replyTo: ActorRef[QueryResult]) extends Command
+  final case class QueryResult(wsid: String, dataType: DataType, func: Function, readings: Int, value: Vector[TimeWindow]) extends CborSerializable
+
+  // small domain model for queriying and storing weather data
+  sealed trait DataType
+  object DataType {
+    /** Temperature in celcius */
+    case object Temperature extends DataType
+    case object Dewpoint extends DataType
+    case object Pressure extends DataType
+    case object WindDirection extends DataType
+    case object OneHourPrecip extends DataType
+    case object SixHourPrecip extends DataType
+    case object SkyCondition extends DataType
+    val All: Set[DataType] = Set(Temperature, Dewpoint, Pressure, WindDirection, OneHourPrecip, SixHourPrecip, SkyCondition)
+  }
+
+  sealed trait Function
+  object Function {
+    case object HighLow extends Function
+    case object Average extends Function
+    case object Current extends Function
+    val All: Set[Function] = Set(HighLow, Average, Current)
+  }
+
+  /**
+   * Simplified weather measurement data, actual weather data event comprises many more data points.
+   *
+   * @param eventTime unix timestamp when collected
+   * @param dataType type of data
+   * @param value data point value
+   */
+  final case class Data(eventTime: Long, dataType: DataType, value: Double)
+  final case class TimeWindow(start: Long, end: Long, value: Double)
```

Comment:
a lot if not all these types lived in the http endpoint logic, making a circular dependency between the entity and the endpoint. They belong here since the entity could be used completely without the http routes etc.